pkg/segment/query/processor/searcher.go

```diff
@@ -387,13 +424,66 @@ func (s *Searcher) initializeQSRs() error {
 	return nil
 }
 
-func (s *Searcher) getBlocks() ([]*block, error) {
-	qsrs, err := query.GetSortedQSRs(s.queryInfo, s.startTime, s.querySummary)
-	if err != nil {
-		log.Errorf("qid=%v, searcher.getBlocks: failed to get sorted QSRs: %v", s.qid, err)
-		return nil, err
+func (s *Searcher) getFilteredBlocks(blocks []*block) []*block {
+	filteredBlocks := make([]*block, 0)
+
+	for _, block := range blocks {
+		if s.processedBlocks[block.parentQSR.GetSegKey()] == nil {
+			s.processedBlocks[block.parentQSR.GetSegKey()] = make(map[uint16]struct{})
+		}
+		_, processed := s.processedBlocks[block.parentQSR.GetSegKey()][block.BlkNum]
+		if processed {
+			continue
+		}
+		if block.HighTs >= s.minTimestamp {
+			filteredBlocks = append(filteredBlocks, block)
+			s.processedBlocks[block.parentQSR.GetSegKey()][block.BlkNum] = struct{}{}
+		}
+	}
+
+	return filteredBlocks
+}
+
+func (s *Searcher) getQSRSToProcess() []*query.QuerySegmentRequest {
+	qsrs := make([]*query.QuerySegmentRequest, 0)
+
+	prevMinTimestamp := s.minTimestamp
+
+	s.segmentIdx += segmentBatchOffset
+	if s.segmentIdx >= len(s.qsrs) {
+		s.segmentIdx = len(s.qsrs)
+		s.minTimestamp = 0
+		s.gotAllSegments = true
```

Comment:
Is this necessary? `s.minTimestamp = 0`

Reply:
addressed in the latest commit. Removed this and cleaned up some code.